File: notifier.py
import logging
import discord
from datetime import datetime, timezone, timedelta

from ball_parser import Lang, format_balls
from settings import SettingsRepository

JST = timezone(timedelta(hours=9))

logger = logging.getLogger(__name__)

_MONTHS_EN = [
    "January", "February", "March", "April", "May", "June",
    "July", "August", "September", "October", "November", "December",
]


def _build_message(ball_ids: list[int], now: datetime, lang: Lang = Lang.JA) -> str:
    if lang == Lang.EN:
        header = f"# {_MONTHS_EN[now.month - 1]} {now.day}"
    else:
        header = f"# {now.month}/{now.day}"

    lines = [header]
    for ball in format_balls(ball_ids, lang):
        lines.append(f"## {ball.name}")
        if ball.description:
            lines.append(ball.description)
    return "\n".join(lines)


class Notifier:
    def __init__(self, client: discord.Client, settings: SettingsRepository) -> None:
        self._client = client
        self._settings = settings

    async def notify(self, ball_ids: list[int]) -> None:
        now = datetime.now(JST)
        for cfg in self._settings.get_all_channels():
            message = _build_message(ball_ids, now, cfg.lang)

            channel = self._client.get_channel(cfg.channel_id)
            if channel is None:
                try:
                    channel = await self._client.fetch_channel(cfg.channel_id)
                except discord.NotFound:
                    logger.warning("Channel %d not found (guild=%s)", cfg.channel_id, cfg.guild_id)
                    continue
                except discord.Forbidden:
                    logger.warning(
                        "No permission to access channel %d (guild=%s)", cfg.channel_id, cfg.guild_id
                    )
                    continue

            role_id = self._settings.get_mention_role(cfg.guild_id)
            content = f"<@&{role_id}>\n{message}" if role_id is not None else message
            await channel.send(content)
            logger.info(
                "Sent special balls %s to channel %d (guild=%s, lang=%s)",
                ball_ids, cfg.channel_id, cfg.guild_id, cfg.lang,
            )

File: internal/bot/notifier.go
package bot

import (
	"errors"
	"fmt"
	"log"
	"strings"
	"sync"
	"time"

	"github.com/bwmarrin/discordgo"

	"github.com/njm2360/dekapu-daily-sherbi-bot/internal/ball"
	"github.com/njm2360/dekapu-daily-sherbi-bot/internal/balltext"
	"github.com/njm2360/dekapu-daily-sherbi-bot/internal/detector"
	"github.com/njm2360/dekapu-daily-sherbi-bot/internal/language"
	"github.com/njm2360/dekapu-daily-sherbi-bot/internal/settings"
)

var monthsEN = [...]string{
	"January", "February", "March", "April", "May", "June",
	"July", "August", "September", "October", "November", "December",
}

func dateHeader(now time.Time, lang language.Lang) string {
	if lang == language.LangEN {
		return fmt.Sprintf("%s %d", monthsEN[int(now.Month())-1], now.Day())
	}
	return fmt.Sprintf("%d/%d", int(now.Month()), now.Day())
}

func buildMessage(kind detector.Kind, ballIDs []int, now time.Time, lang language.Lang) string {
	var sb strings.Builder
	sb.WriteString("# ")
	sb.WriteString(dateHeader(now, lang))
	if kind == detector.SeedUpdate {
		if lang == language.LangEN {
			sb.WriteString(" (Daily updated by seed change)")
		} else {
			sb.WriteString(" (シード更新でデイリーが変わったよ)")
		}
	}

	for _, info := range balltext.Format(ballIDs, lang) {
		sb.WriteString("\n## ")
		sb.WriteString(info.Name)
		if info.Description != "" {
			sb.WriteString("\n")
			sb.WriteString(info.Description)
		}
	}
	return sb.String()
}

type Notifier struct {
	session *discordgo.Session
	repo    *settings.Repository
}

func NewNotifier(s *discordgo.Session, r *settings.Repository) *Notifier {
	return &Notifier{session: s, repo: r}
}

func (n *Notifier) Notify(kind detector.Kind, daily ball.Daily) {
	date := daily.Date()

	channels, err := n.repo.AllChannels()
	if err != nil {
		log.Printf("Notify: AllChannels: %v", err)
		return
	}

	var wg sync.WaitGroup
	for _, cfg := range channels {
		wg.Add(1)
		go func(cfg settings.ChannelConfig) {
			defer wg.Done()

			message := buildMessage(kind, daily.BallIDs, date, cfg.Lang)
			channelID := cfg.ChannelID

			roleID, hasRole, err := n.repo.GetMentionRole(cfg.GuildID)
			if err != nil {
				log.Printf("Notify: GetMentionRole guild=%s: %v", cfg.GuildID, err)
			}
			content := message
			if hasRole {
				content = fmt.Sprintf("<@&%s>\n%s", roleID, message)
			}

			if _, err := n.session.ChannelMessageSend(channelID, content); err != nil {
				var rest *discordgo.RESTError
				if errors.As(err, &rest) && rest.Response != nil {
					switch rest.Response.StatusCode {
					case 404:
						log.Printf("Channel %s not found (guild=%s)", channelID, cfg.GuildID)
						return
					case 403:
						log.Printf("No permission to access channel %s (guild=%s)", channelID, cfg.GuildID)
						return
					}
				}
				log.Printf("ChannelMessageSend failed (channel=%s, guild=%s): %v",
					channelID, cfg.GuildID, err)
				return
			}
			log.Printf("Sent special balls %v (day=%d) to channel %s (guild=%s, lang=%s)",
				daily.BallIDs, daily.DayNumber, channelID, cfg.GuildID, cfg.Lang)
		}(cfg)
	}
	wg.Wait()
}
